lib/stencil-pull.utils.js

```diff
@@ -1,54 +1,79 @@
 const themeApiClient = require('./theme-api-client');
-const tmp = require('tmp-promise');
-const { extractZipFiles } = require('./archiveManager');
-const { fetchFile } = require('./utils/networkUtils');
+const fs = require('fs');
+const _ = require('lodash');
 
 const utils = {};
 
 module.exports = utils;
 
-utils.selectActiveTheme = (options, callback) => {
-    const [activeTheme] = options.themes.filter(theme => theme.is_active).map(theme => theme.uuid);
+utils.getChannelActiveTheme = async options => {
+    const { config: { accessToken }, apiHost, storeHash, channelId } = options;
 
-    callback(null, Object.assign({}, options, { activeTheme }));
+    const activeTheme = await themeApiClient.getChannelActiveTheme({ accessToken, apiHost, storeHash, channelId });
+
+    console.log('ok'.green + ` -- Fetched theme details for channel ${channelId}`);
+
+    return { ...options, activeTheme };
 };
 
-utils.startThemeDownloadJob = async options => {
-    const { config: { accessToken }, apiHost, activeTheme, storeHash } = options;
+utils.getThemeConfiguration = async options => {
+    const { config: { accessToken }, apiHost, storeHash, activeTheme, saved } = options;
+
+    const themeId = activeTheme.active_theme_uuid;
+
+    const configurationId = saved ? activeTheme.saved_theme_configuration_uuid
+        : activeTheme.active_theme_configuration_uuid;
 
-    const { jobId } = await themeApiClient.downloadTheme({
-        accessToken: accessToken,
-        themeId: activeTheme,
-        apiHost,
-        storeHash,
-    });
+    const remoteThemeConfiguration = await themeApiClient.getThemeConfiguration({ accessToken, apiHost, storeHash,
+                                                                                  themeId, configurationId });
 
-    return {
-        ...options,
-        jobId,
-    };
+    console.log('ok'.green + ` -- Fetched ${saved ? 'saved' : 'active'} configuration`);
+
+    return { ...options, remoteThemeConfiguration };
 };
 
-utils.downloadThemeConfig = async options => {
-    const { path: tempThemePath, cleanup } = await tmp.file();
+utils.mergeThemeConfiguration = async options => {
+    const { remoteThemeConfiguration } = options;
 
-    try {
-        await fetchFile(options.downloadUrl, tempThemePath);
-    } catch (err) {
-        throw new Error(`Unable to download theme config from ${options.downloadUrl}: ${err.message}`);
-    }
+    let rawConfig = fs.readFileSync('config.json');
+    let parsedConfig = JSON.parse(rawConfig);
+    let diffDetected = false;
 
-    console.log('ok'.green + ' -- Theme files downloaded');
-    console.log('ok'.green + ' -- Extracting theme config');
+    // For any keys the remote configuration has in common with the local configuration,
+    // overwrite the local configuration if the remote configuration differs
+    for (const [key, value] of Object.entries(remoteThemeConfiguration.settings)) {
```

Comment:
@bookernath why do we take into account only the `config.settings` field?
Can't other fields be changed? (e.g. `config.variations.settings`).

Reply:
Can we just overwrite the file without checking the difference (it will be faster and simpler)?

Reply:
@MaxGenash thanks for your questions/comments.

The config in the remote may actually have different keys, so we can't just save it directly.

For example, we may have this case...

Remote configuration
```
{
  "settings": {
      "a": 1,
      "b":  8
  }
}
```

Local configuration
```
{
  "settings": {
      "b":  2,
      "c": 3
  }
}
```

In which case the correct "merged" configuration will prefer the remote configuration for any keys in common, ignoring keys that don't exist locally. So we'd want:

Merged configuration
```
{
  "settings": {
      "b":  8,
      "c": 3
  }
}
```

Couple of assumptions

1. If a key is missing from the local configuration, it's not used in the local theme, so it's not needed
2. If a key is in both configurations but has a different value, we want to update the local theme with the remote configuration so that `stencil push` doesn't overwrite the remote configuration (this is the point of `stencil pull`)
3. If a key is missing from the remote configuration that is present in the local configuration, it's probably a new key that was added - so the local theme needs it, but the remote theme doesn't
4. If we have a key in common with a different data type, this is a breaking change and the user needs to know about it via an error

Reply:
Regarding *variations*, these just act as presets for settings which ultimately end up in config.json. But if you make a change in Page Builder which saves an explicit setting for one of these keys, that value takes precedence. So correct behavior for `stencil pull` is to make `config.json` match the remote configuration (ignoring variations) to make sure storefront behavior is not changed by the next deployment via `stencil pull`